File: app/src/androidTest/java/org/tasks/makers/CaldavTaskMaker.java
package org.tasks.makers;

import static com.natpryce.makeiteasy.Property.newProperty;
import static org.tasks.makers.Maker.make;

import com.natpryce.makeiteasy.Instantiator;
import com.natpryce.makeiteasy.Property;
import com.natpryce.makeiteasy.PropertyValue;
import org.tasks.data.CaldavTask;

public class CaldavTaskMaker {

  public static final Property<CaldavTask, String> CALENDAR = newProperty();
  public static final Property<CaldavTask, Long> TASK = newProperty();
  public static final Property<CaldavTask, String> REMOTE_ID = newProperty();
  public static final Property<CaldavTask, String> REMOTE_PARENT = newProperty();

  private static final Instantiator<CaldavTask> instantiator =
      lookup -> {
        CaldavTask task =
            new CaldavTask(lookup.valueOf(TASK, 1L), lookup.valueOf(CALENDAR, "calendar"));
        task.setRemoteId(lookup.valueOf(REMOTE_ID, task.getRemoteId()));
        task.setRemoteParent(lookup.valueOf(REMOTE_PARENT, (String) null));
        return task;
      };

  @SafeVarargs
  public static CaldavTask newCaldavTask(PropertyValue<? super CaldavTask, ?>... properties) {
    return make(instantiator, properties);
  }
}

File: app/src/androidTest/java/org/tasks/makers/CaldavTaskMaker.kt
package org.tasks.makers

import com.natpryce.makeiteasy.Instantiator
import com.natpryce.makeiteasy.Property
import com.natpryce.makeiteasy.Property.newProperty
import com.natpryce.makeiteasy.PropertyValue
import org.tasks.data.CaldavTask
import org.tasks.makers.Maker.make

object CaldavTaskMaker {
    @JvmField val CALENDAR: Property<CaldavTask, String> = newProperty()
    @JvmField val TASK: Property<CaldavTask, Long> = newProperty()
    @JvmField val REMOTE_ID: Property<CaldavTask, String?> = newProperty()
    @JvmField val REMOTE_PARENT: Property<CaldavTask, String?> = newProperty()

    private val instantiator = Instantiator<CaldavTask> {
        val task = CaldavTask(it.valueOf(TASK, 1L), it.valueOf(CALENDAR, "calendar"))
        task.remoteId = it.valueOf(REMOTE_ID, task.remoteId)
        task.remoteParent = it.valueOf(REMOTE_PARENT, null as String?)
        task
    }

    @JvmStatic
    fun newCaldavTask(vararg properties: PropertyValue<in CaldavTask?, *>): CaldavTask {
        return make(instantiator, *properties)
    }
}